File: queue.js
class Node {
    constructor(value) {
        this.value = value;
        this.next = null;
    }
}

class Queue{
    constructor() {
        this.first = null;
        this.last = null;
        this.length = 0;
    }
    peek() {
        return this.first;
    }
    enqueue(value) {
        const newNode = new Node(value);
        if(this.length === 0) {
            this.first = newNode;
            this.last = newNode
            this.length++;
        } else {
            this.last.next = newNode;
            this.last = newNode;
            this.length++;
        }
        return this;
    }
    dequeue() {
        if(!this.first) {
            return null;
        }
        if(this.first === this.last) {
            this.last = null;
        }
        this.first = this.first.next;
        this.length--;
        console.log(this);
        return this;
    }
}
const myQueue = new Queue();
myQueue.enqueue('Roj');
myQueue.enqueue('Cali');
myQueue.enqueue('Camille');
myQueue.dequeue();
myQueue.dequeue();
myQueue.dequeue();

File: JAVA_VERSION/DATA-STRUCTURE_ALGORITHM/src/datastructure/queue/Queue.java
package datastructure.queue;

public class Queue {
    private Node first;
    private Node last;
    private int length;

    class Node {
        private int value;
        private Node next;
        Node(int value) {
            this.value = value;
        }
    }

    Queue(int value) {
        Node newNode = new Node(value);
        this.first = newNode;
        this.last = newNode;
        this.length++;
    }

    public void enqueue(int value) {
        Node newNode = new Node(value);
        if(this.length == 0) {
            this.first = newNode;
            this.last = newNode;
        } else {
            this.last.next = newNode;
            this.last = newNode;
        }
        this.length++;
    }

    public Node dequeue() {
        Node temp = this.first;
        if(this.length == 0) {
            return null;
        }
        else if(length == 1) {
            this.first = null;
            this.last = null;
        } else {
            this.first = temp.next;
            temp.next = null;
        }
        this.length--;
        return temp;

    }

    public void printQueue() {
        Node temp = this.first;
        while(temp != null) {
            System.out.println(temp.value);
            temp = temp.next;
        }
    }

    public void getFirst() {
        System.out.println("First: " + this.first.value);
    }

    public void getLast() {
        System.out.println("Last: " + this.last.value);
    }

    public void getLength() {
        System.out.println("Length: " + this.length);
    }
}
